File: android/app/src/main/java/com/ayuvo/health/actions/ActionMath.kt
package com.ayuvo.health.actions

import com.ayuvo.health.data.metrics.MetricsReference
import com.ayuvo.health.data.metrics.WeekStart
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.floor

/** Half-open `[fromMs, toMs)` in the user's zone. */
data class ActionDateRange(val fromMs: Long, val toMs: Long) {
    companion object {
        /** `resolve_date_range` (shared/actions/test-vectors/date_ranges.json). */
        fun resolve(preset: String, nowMs: Long, zone: ZoneId, weekStart: WeekStart): ActionDateRange {
            val d = MetricsReference.localDateOf(nowMs, zone)
            val (a, b) = when (preset) {
                "today" -> d to d.plusDays(1)
                "yesterday" -> d.minusDays(1) to d
                "this_week", "last_week" -> {
                    val ws = MetricsReference.weekStartOf(d, weekStart)
                    if (preset == "this_week") ws to ws.plusDays(7) else ws.minusDays(7) to ws
                }
                "last_7_days" -> d.minusDays(6) to d.plusDays(1)
                "last_30_days" -> d.minusDays(29) to d.plusDays(1)
                "this_month", "last_month" -> {
                    val first = LocalDate.of(d.year, d.month, 1)
                    if (preset == "this_month") first to first.plusMonths(1) else first.minusMonths(1) to first
                }
                else -> throw IllegalArgumentException("bad date range $preset")
            }
            return ActionDateRange(MetricsReference.localMidnight(a, zone), MetricsReference.localMidnight(b, zone))
        }
    }

    operator fun contains(ms: Long): Boolean = ms in fromMs until toMs
}

/**
 * The small calculations action outputs share with iOS (shared/actions/test-vectors/compute.json).
 * Rounding is half up by `floor(x * 10^d + 0.5) / 10^d` on every platform.
 */
object ActionMath {
    fun roundTo(x: Double, decimals: Int): Double {
        var scale = 1.0
        repeat(decimals) { scale *= 10.0 }
        return floor(x * scale + 0.5) / scale
    }

    fun percent(part: Double, whole: Double): Int = floor(part * 100.0 / whole + 0.5).toInt()

    fun toCanonical(catalog: ActionCatalog, family: String, unit: String, value: Double): Double =
        value * catalog.units.getValue(family).factors.getValue(unit)

    fun convert(catalog: ActionCatalog, family: String, from: String, to: String, value: Double): Double {
        val factors = catalog.units.getValue(family).factors
        return roundTo(value * factors.getValue(from) / factors.getValue(to), 3)
    }

    data class Sample(val tMs: Long, val value: Double?)

    /** `aggregate`: `value` is null without samples (0 for count); average is rounded to 2 decimals. */
    fun aggregate(entries: List<Sample>, aggregation: String): Map<String, Any?> {
        val vals = entries.filter { it.value != null }
        val n = vals.size
        if (aggregation == "count") return linkedMapOf("value" to n.toLong(), "count" to n.toLong())
        if (n == 0) return linkedMapOf("value" to null, "count" to 0L)
        val value: Double = when (aggregation) {
            "latest" -> {
                var best = vals[0]
                for (s in vals.drop(1)) if (s.tMs >= best.tMs) best = s
                best.value!!
            }
            "sum" -> {
                var total = 0.0
                for (s in vals) total += s.value!!
                total
            }
            "average" -> {
                var total = 0.0
                for (s in vals) total += s.value!!
                roundTo(total / n, 2)
            }
            "min" -> vals.minOf { it.value!! }
            "max" -> vals.maxOf { it.value!! }
            else -> throw IllegalArgumentException("bad aggregation $aggregation")
        }
        return linkedMapOf("value" to value, "count" to n.toLong())
    }

    fun waterStatus(intakeMl: Long, goalMl: Long?): Map<String, Any?> {
        val hasGoal = goalMl != null && goalMl > 0
        return linkedMapOf(
            "intake_ml" to intakeMl,
            "goal_ml" to if (hasGoal) goalMl else null,
            "remaining_ml" to if (hasGoal) maxOf(0L, goalMl!! - intakeMl) else null,
            "percent" to if (hasGoal) percent(intakeMl.toDouble(), goalMl!!.toDouble()).toLong() else null
        )
    }

    fun progress(value: Double, target: Double?): Map<String, Any?> {
        val has = target != null && target > 0
        return linkedMapOf(
            "value" to roundTo(value, 1),
            "target" to if (has) roundTo(target!!, 1) else null,
            "remaining" to if (has) roundTo(maxOf(0.0, target!! - value), 1) else null,
            "percent" to if (has) percent(value, target!!).toLong() else null
        )
    }

    fun bmi(weightKg: Double?, heightCm: Double?): Double? {
        if (weightKg == null || heightCm == null || weightKg <= 0 || heightCm <= 0) return null
        val m = heightCm / 100.0
        return roundTo(weightKg / (m * m), 1)
    }

    data class WeightPoint(val tMs: Long, val kg: Double)

    fun weightChange(entries: List<WeightPoint>): Map<String, Any?> {
        val ordered = entries.sortedBy { it.tMs }
        if (ordered.isEmpty()) return linkedMapOf("first_kg" to null, "last_kg" to null, "change_kg" to null, "count" to 0L)
        val first = ordered.first().kg
        val last = ordered.last().kg
        return linkedMapOf(
            "first_kg" to roundTo(first, 1),
            "last_kg" to roundTo(last, 1),
            "change_kg" to roundTo(last - first, 1),
            "count" to ordered.size.toLong()
        )
    }

    data class SetWork(val weightKg: Double, val reps: Int)

    fun setVolume(sets: List<SetWork>): Map<String, Any?> {
        var reps = 0L
        var volume = 0.0
        for (s in sets) {
            reps += s.reps
            volume += s.weightKg * s.reps
        }
        return linkedMapOf("sets" to sets.size.toLong(), "reps" to reps, "volume_kg" to roundTo(volume, 1))
    }

    fun fastingProgress(startedMs: Long, goalMinutes: Long, nowMs: Long): Map<String, Any?> {
        val elapsed = maxOf(0L, Math.floorDiv(nowMs - startedMs, 1000L))
        val goal = goalMinutes * 60
        return linkedMapOf(
            "elapsed_s" to elapsed,
            "goal_s" to goal,
            "remaining_s" to maxOf(0L, goal - elapsed),
            "percent" to if (goal > 0) percent(elapsed.toDouble(), goal.toDouble()).toLong() else null,
            "reached" to (goal > 0 && elapsed >= goal)
        )
    }
}

File: ios/calorietracker/Actions/ActionMath.swift
import Foundation

/// Date ranges, unit conversion and the small calculations action outputs share with Android —
/// Swift port of `resolve_date_range` and `compute` in `scripts/actions_reference.py`
/// (vectors: `shared/actions/test-vectors/date_ranges.json`, `compute.json`).
nonisolated enum ActionMath {
    typealias Zone = MetricsReference.Zone
    typealias LocalDay = MetricsReference.LocalDay

    // MARK: - Rounding (half up, same formula on every platform)

    static func roundTo(_ x: Double, _ decimals: Int) -> Double {
        let scale = pow(10, Double(decimals))
        return floor(x * scale + 0.5) / scale
    }

    static func percent(_ part: Double, _ whole: Double) -> Int {
        Int(floor(part * 100.0 / whole + 0.5))
    }

    // MARK: - Date ranges: half-open [fromMs, toMs) in the user's time zone

    struct DateRange: Equatable, Sendable {
        let fromMs: Int64
        let toMs: Int64
        let firstDay: LocalDay
        let dayCount: Int

        var from: Date { Date(timeIntervalSince1970: Double(fromMs) / 1000) }
        var to: Date { Date(timeIntervalSince1970: Double(toMs) / 1000) }
        var isSingleDay: Bool { dayCount == 1 }

        func contains(_ date: Date) -> Bool {
            let ms = Int64((date.timeIntervalSince1970 * 1000).rounded())
            return ms >= fromMs && ms < toMs
        }

        func contains(ms: Int64) -> Bool { ms >= fromMs && ms < toMs }
    }

    static func resolveDateRange(_ preset: String, nowMs: Int64, zone: Zone, weekStart: MetricsReference.WeekStart) -> DateRange? {
        let d = zone.day(of: nowMs)
        let days: (LocalDay, LocalDay)
        switch preset {
        case "today": days = (d, d.adding(days: 1))
        case "yesterday": days = (d.adding(days: -1), d)
        case "this_week", "last_week":
            let ws = MetricsReference.weekStartOf(d, weekStart)
            days = preset == "this_week" ? (ws, ws.adding(days: 7)) : (ws.adding(days: -7), ws)
        case "last_7_days": days = (d.adding(days: -6), d.adding(days: 1))
        case "last_30_days": days = (d.adding(days: -29), d.adding(days: 1))
        case "this_month", "last_month":
            let first = LocalDay(year: d.year, month: d.month, day: 1)
            days = preset == "this_month" ? (first, first.addingMonths(1)) : (first.addingMonths(-1), first)
        default: return nil
        }
        let (a, b) = days
        return DateRange(fromMs: zone.midnight(a), toMs: zone.midnight(b), firstDay: a, dayCount: b.ordinal - a.ordinal)
    }

    // MARK: - Units

    static func toCanonical(_ value: Double, family: ActionCatalog.UnitFamily, unit: String?) -> Double {
        value * (family.factors[unit ?? family.canonical] ?? 1)
    }

    static func convert(_ value: Double, family: ActionCatalog.UnitFamily, from: String, to: String) -> Double {
        roundTo(value * (family.factors[from] ?? 1) / (family.factors[to] ?? 1), 3)
    }

    // MARK: - Calculations

    struct Sample: Sendable, Equatable {
        let tMs: Int64
        let value: Double?
    }

    struct Aggregate: Equatable, Sendable {
        let value: Double?
        let count: Int
    }

    static func aggregate(_ entries: [Sample], _ aggregation: String) -> Aggregate {
        let values = entries.compactMap { e in e.value.map { (e.tMs, $0) } }
        let n = values.count
        if aggregation == "count" { return Aggregate(value: Double(n), count: n) }
        guard let head = values.first else { return Aggregate(value: nil, count: 0) }
        switch aggregation {
        case "latest":
            var best = head
            for item in values.dropFirst() where item.0 >= best.0 { best = item }
            return Aggregate(value: best.1, count: n)
        case "sum":
            var total = 0.0
            for item in values { total += item.1 }
            return Aggregate(value: total, count: n)
        case "average":
            var total = 0.0
            for item in values { total += item.1 }
            return Aggregate(value: roundTo(total / Double(n), 2), count: n)
        case "min":
            return Aggregate(value: values.map(\.1).min(), count: n)
        case "max":
            return Aggregate(value: values.map(\.1).max(), count: n)
        default:
            return Aggregate(value: nil, count: n)
        }
    }

    struct WaterStatus: Equatable, Sendable {
        let intakeMl: Int
        let goalMl: Int?
        let remainingMl: Int?
        let percent: Int?
    }

    static func waterStatus(intakeMl: Int, goalMl: Int?) -> WaterStatus {
        guard let goalMl, goalMl > 0 else { return WaterStatus(intakeMl: intakeMl, goalMl: nil, remainingMl: nil, percent: nil) }
        return WaterStatus(intakeMl: intakeMl, goalMl: goalMl, remainingMl: max(0, goalMl - intakeMl),
                           percent: percent(Double(intakeMl), Double(goalMl)))
    }

    struct Progress: Equatable, Sendable {
        let value: Double
        let target: Double?
        let remaining: Double?
        let percent: Int?
    }

    static func progress(value: Double, target: Double?) -> Progress {
        guard let target, target > 0 else { return Progress(value: roundTo(value, 1), target: nil, remaining: nil, percent: nil) }
        return Progress(value: roundTo(value, 1), target: roundTo(target, 1),
                        remaining: roundTo(max(0.0, target - value), 1), percent: percent(value, target))
    }

    static func bmi(weightKg: Double?, heightCm: Double?) -> Double? {
        guard let weightKg, let heightCm, weightKg > 0, heightCm > 0 else { return nil }
        let m = heightCm / 100.0
        return roundTo(weightKg / (m * m), 1)
    }

    struct WeightChange: Equatable, Sendable {
        let firstKg: Double?
        let lastKg: Double?
        let changeKg: Double?
        let count: Int
    }

    static func weightChange(_ entries: [(tMs: Int64, kg: Double)]) -> WeightChange {
        let ordered = entries.enumerated().sorted { ($0.element.tMs, $0.offset) < ($1.element.tMs, $1.offset) }.map(\.element)
        guard let first = ordered.first, let last = ordered.last else {
            return WeightChange(firstKg: nil, lastKg: nil, changeKg: nil, count: 0)
        }
        return WeightChange(firstKg: roundTo(first.kg, 1), lastKg: roundTo(last.kg, 1),
                            changeKg: roundTo(last.kg - first.kg, 1), count: ordered.count)
    }

    struct SetVolume: Equatable, Sendable {
        let sets: Int
        let reps: Int
        let volumeKg: Double
    }

    static func setVolume(_ sets: [(weightKg: Double, reps: Int)]) -> SetVolume {
        var reps = 0
        var volume = 0.0
        for set in sets {
            reps += set.reps
            volume += set.weightKg * Double(set.reps)
        }
        return SetVolume(sets: sets.count, reps: reps, volumeKg: roundTo(volume, 1))
    }

    struct FastingProgress: Equatable, Sendable {
        let elapsedS: Int64
        let goalS: Int64
        let remainingS: Int64
        let percent: Int?
        let reached: Bool
    }

    static func fastingProgress(startedMs: Int64, goalMinutes: Int, nowMs: Int64) -> FastingProgress {
        let elapsed = max(0, (nowMs - startedMs) / 1000)
        let goal = Int64(goalMinutes) * 60
        return FastingProgress(elapsedS: elapsed, goalS: goal, remainingS: max(0, goal - elapsed),
                               percent: goal > 0 ? percent(Double(elapsed), Double(goal)) : nil,
                               reached: goal > 0 && elapsed >= goal)
    }

    // MARK: - Vector adapter (`compute` in the reference)

    static func compute(catalog: ActionCatalog, op: String, args: [String: Any]) -> [String: Any]? {
        func double(_ key: String) -> Double? { (args[key] as? NSNumber)?.doubleValue }
        func null(_ value: Any?) -> Any { value ?? NSNull() }
        switch op {
        case "convert":
            guard let familyName = args["family"] as? String, let family = catalog.units[familyName],
                  let from = args["from"] as? String, let to = args["to"] as? String, let value = double("value") else { return nil }
            return ["value": convert(value, family: family, from: from, to: to)]
        case "aggregate":
            let entries = (args["entries"] as? [[String: Any]] ?? []).map {
                Sample(tMs: ($0["t_ms"] as? NSNumber)?.int64Value ?? 0, value: ($0["value"] as? NSNumber)?.doubleValue)
            }
            let result = aggregate(entries, args["aggregation"] as? String ?? "")
            return ["value": null(result.value), "count": result.count]
        case "water_status":
            let status = waterStatus(intakeMl: (args["intake_ml"] as? NSNumber)?.intValue ?? 0,
                                     goalMl: (args["goal_ml"] as? NSNumber)?.intValue)
            return ["intake_ml": status.intakeMl, "goal_ml": null(status.goalMl), "remaining_ml": null(status.remainingMl),
                    "percent": null(status.percent)]
        case "progress":
            let p = progress(value: double("value") ?? 0, target: double("target"))
            return ["value": p.value, "target": null(p.target), "remaining": null(p.remaining), "percent": null(p.percent)]
        case "bmi":
            return ["bmi": null(bmi(weightKg: double("weight_kg"), heightCm: double("height_cm")))]
        case "weight_change":
            let entries = (args["entries"] as? [[String: Any]] ?? []).map {
                (tMs: ($0["t_ms"] as? NSNumber)?.int64Value ?? 0, kg: ($0["kg"] as? NSNumber)?.doubleValue ?? 0)
            }
            let c = weightChange(entries)
            return ["first_kg": null(c.firstKg), "last_kg": null(c.lastKg), "change_kg": null(c.changeKg), "count": c.count]
        case "set_volume":
            let sets = (args["sets"] as? [[String: Any]] ?? []).map {
                (weightKg: ($0["weight_kg"] as? NSNumber)?.doubleValue ?? 0, reps: ($0["reps"] as? NSNumber)?.intValue ?? 0)
            }
            let v = setVolume(sets)
            return ["sets": v.sets, "reps": v.reps, "volume_kg": v.volumeKg]
        case "fasting_progress":
            let p = fastingProgress(startedMs: (args["started_ms"] as? NSNumber)?.int64Value ?? 0,
                                    goalMinutes: (args["goal_minutes"] as? NSNumber)?.intValue ?? 0,
                                    nowMs: (args["now_ms"] as? NSNumber)?.int64Value ?? 0)
            return ["elapsed_s": p.elapsedS, "goal_s": p.goalS, "remaining_s": p.remainingS, "percent": null(p.percent),
                    "reached": p.reached]
        default:
            return nil
        }
    }
}
